mbrl/util/pybullet.py

```diff
@@ -0,0 +1,249 @@
+        reward_fn = None
+
+    if cfg.seed is not None:
+        env.seed(cfg.seed)
+        env.observation_space.seed(cfg.seed + 1)
+        env.action_space.seed(cfg.seed + 2)
+
+    return env, term_fn, reward_fn
+
+def make_env_from_str(env_name: str) -> gym.Env:
+    """Creates a new environment from its string description.
+
+    Args:
+        env_name (str): the string description of the environment. Valid options are:
+          - "gym___<env_name>": a Gym environment (e.g., "gym___HalfCheetah-v2").
+
+    Returns:
+        (gym.Env): the created environment.
+    """
+    if "gym___" in env_name:
+        env = gym.make(env_name.split("___")[1])
+    else:
+        raise ValueError("Invalid environment string.")
+
+    return env
+
+def _is_pybullet_gym_env(env: gym.wrappers.TimeLimit) -> bool:
+    # I don't know how to do this yet
+    return True
+
+class freeze_pybullet_env:
+    """Provides a context to freeze a Pybullet environment.
+
+    This context allows the user to manipulate the state of a PyBullet environment and return it
+    to its original state upon exiting the context.
+
+    Works with pybulletgym environments
+
+    Example usage:
+
+    .. code-block:: python
+
+       env = gym.make("HalfCheetahPyBulletEnv-v0")
+       env.reset()
+       action = env.action_space.sample()
+       # o1_expected, *_ = env.step(action)
+       with freeze_pybullet_env(env):
+           step_the_env_a_bunch_of_times()
+       o1, *_ = env.step(action) # o1 will be equal to what o1_expected would have been
+
+    Args:
+        env (:class:`gym.wrappers.TimeLimit`): the environment to freeze.
+    """
+
+    def __init__(self, env: gym.wrappers.TimeLimit):
+        self._env = env
+        self._init_state: np.ndarray = None
+        self._elapsed_steps = 0
+        self._step_count = 0
+
+        if _is_pybullet_gym_env(env):
+            self._enter_method = self._enter_pybullet_gym
+            self._exit_method = self._exit_pybullet_gym
+        else:
+            raise RuntimeError("Tried to freeze an unsupported environment.")
+
+    def _enter_pybullet_gym(self):
+        # For now, the accepted envs are limited to ease implementation and testing
+        from pybulletgym.envs.roboschool.robots.locomotors.walker_base import  WalkerBase as RSWalkerBase
+        from pybulletgym.envs.mujoco.robots.locomotors.walker_base import WalkerBase as MJWalkerBase
+        env = self._env.env
+        robot = env.robot
+        assert isinstance(robot, (RSWalkerBase, MJWalkerBase))
+        self.state_id = env._p.saveState()
+        self.ground_ids = env.ground_ids
+        self.potential = env.potential
+        self.reward = float(env.reward)
+        robot_keys = [("body_rpy", tuple), ("body_xyz", tuple), ("feet_contact", np.copy), ("initial_z", float), ("joint_speeds", np.copy), ("joints_at_limit", int), ("walk_target_dist", float), ("walk_target_theta", float), ("walk_target_x", float), ("walk_target_y", float)]
```

Comment:
Looks like you are missing the pre-commit hooks that make it easier to adhere to our code format, and l just noticed that I haven't added it to our dev requirements.txt file. You can install it as `pip install pre-commit`. Once installed, it will run `black`, `flake8`, `mypy`, and `isort` before you do any commit and block the commit if something is amiss.

Reply:
Got it, I have installed that.

Reply:
@luisenp I'm having a bit of trouble with the pre-commit hook. I am developing in a Conda environment using python 3.8.11 and I ran into an error: 
```
(mbrl-lib) dtch1997@DESKTOP-AR4R24K:~/code/mbrl-lib$ git commit
[WARNING] Unstaged files detected.
[INFO] Stashing unstaged files to /home/dtch1997/.cache/pre-commit/patch1633183394-32437.
[INFO] Installing environment for https://github.com/psf/black.
[INFO] Once installed this environment will be reused.
[INFO] This may take a few minutes...
[INFO] Restored changes from /home/dtch1997/.cache/pre-commit/patch1633183394-32437.
An unexpected error has occurred: CalledProcessError: command: ('/home/dtch1997/anaconda3/envs/mbrl-lib/bin/python', '-mvirtualenv', '/home/dtch1997/.cache/pre-commit/repolvgrel3_/py_env-python3.7', '-p', 'python3.7')
return code: 1
expected return code: 0
stdout:
    RuntimeError: failed to find interpreter for Builtin discover of python_spec='python3.7'

stderr: (none)
Check the log at /home/dtch1997/.cache/pre-commit/pre-commit.log
```
Do I need to install additional python binaries on my device in order to pass the pre-commit checks?

Reply:
I'm actually not sure, I've never gotten this error but I've always been using python 3.7 for this project. I'm guessing it might be related to [this](https://github.com/facebookresearch/mbrl-lib/blob/main/.pre-commit-config.yaml#L6) line, but I would have to look up if there is syntax to support multiple python versions, and it's hard for me to get time for this over the weekend; of course, you are welcome to look into it if you are curious and prefer using your 3.8 env. 

That being said, if you want to make progress in the mean time, try running commit with the option `--no-verify` to skip the hooks while we sort this out.

Reply:
I fixed the above by switching my development environment to python 3.7. 
Now I'm struggling to understand why this happens with black: 
```
(mbrl-lib-py37) dtch1997@DESKTOP-AR4R24K:~/code/mbrl-lib$ git commit
[WARNING] Unstaged files detected.
[INFO] Stashing unstaged files to /home/dtch1997/.cache/pre-commit/patch1633184818-865.
black....................................................................Failed
- hook id: black
- files were modified by this hook
```